### ISSUE TITLE
test/format recover test failed

### ISSUE THREAD

**Issue by sueloverso** [wiredtiger/wiredtiger#1700]
test/format recovery script failed with `WT_NOTFOUND` when recovery is run for the verify step.  
http://build.wiredtiger.com:8080/job/wiredtiger-test-format-recovery/183/console

```
lt-wt: WT_NOTFOUND: item not found
lt-wt: verify(file:wt): No such file or directory
```
This is reproducible when running against a copy of the save directory.

**Comment by michaelcahill** [comment]
Hi @sueloverso, is there a copy of the save directory stashed somewhere?  Are you happy to chase this one or should @keithbostic take a look?

**Comment by sueloverso** [comment]
I am happy to take a look at this and planned to on Wednesday morning.  There is an unpacked copy on tinderbox in ~jenkins/recovery.183/<path>/RUNDIR.SAVE.  Or you can just copy the tar file from ~jenkins/recovery.183/rec183.tgz and unpack it locally.  Running `wt` will reproduce the WT_NOTFOUND error anywhere when run on a copy of that save directory.

**Comment by michaelcahill** [comment]
Hi @sueloverso, I had a quick look at this one.

The following change avoids the `WT_NOTFOUND` error, but I'm not a big fan because it could mask some other case where the LSN can't be found in the log, and skip recovery when it really should fail.

Can `__wt_log_needs_recovery` do explicit tests on the LSN rather than using a log cursor?

```diff
diff --git a/src/log/log.c b/src/log/log.c
--- a/src/log/log.c
+++ b/src/log/log.c
@@ -61,18 +61,19 @@ int
 
 	WT_RET(__wt_curlog_open(session, "log:", NULL, &c));
 	c->set_key(c, ckp_lsn->file, ckp_lsn->offset, 0);
-	WT_ERR(c->search(c));
 
 	/*
 	 * If the checkpoint LSN we're given is the last record, then recovery
-	 * is not needed.
+	 * is not needed.  The search can return WT_NOTFOUND in that case, or
+	 * it can find the last record, and then next will return WT_NOTFOUND.
 	 */
-	if ((ret = c->next(c)) == WT_NOTFOUND) {
+	if ((ret = c->search(c)) == WT_NOTFOUND ||
+	    (ret == 0 && (ret = c->next(c)) == WT_NOTFOUND)) {
 		*rec = 0;
 		ret = 0;
 	}
 
-err:	WT_TRET(c->close(c));
+	WT_TRET(c->close(c));
 	return (ret);
 }
```

**Comment by sueloverso** [comment]
@michaelcahill Looking more closely at this, there is a bug on the side where we generate the checkpoint LSN that we're writing into the metadata.  The problem is that it is hitting an end-of-log edge case.  Here's what's happening:
* We come into `__wt_txn_checkpoint_log` and set `*ckpt_lsn` to `log->alloc_lsn`, in log file N,X.
* `log->alloc_lsn` points to the next LSN we try to use.  In this case X happens to be very close to the end of log file N.
* We write the commit record into the log writing the file's information containing the checkpoint LSN of N,X into the metadata file at the `alloc_lsn` of N,X.
* Logging discovers that the record won't fit and the LSN it is actually written to is N+1,128.
* The metadata now contains an LSN that doesn't exist and the record does exist at something different.

I'm going to think about this.

**Comment by sueloverso** [comment]
Ok, so I've been diving into this and it has led to unexpected paths.  The recovery.sh script executes this command `wt -R -h <dir> verify file:wt`.  That is failing with ENOENT, `file:wt` not found.

First, I came up with a different change to `__wt_log_needs_recovery`.  Essentially, if the code finds that the LSN does not exist, it returns that recovery is needed:
```
diff --git a/src/log/log.c b/src/log/log.c
index f76ec40..320822a 100644
--- a/src/log/log.c
+++ b/src/log/log.c
@@ -61,16 +61,23 @@ __wt_log_needs_recovery(WT_SESSION_IMPL *session, WT_LSN *ckp_lsn, int *rec)
 
 	WT_RET(__wt_curlog_open(session, "log:", NULL, &c));
 	c->set_key(c, ckp_lsn->file, ckp_lsn->offset, 0);
-	WT_ERR(c->search(c));
-
-	/*
-	 * If the checkpoint LSN we're given is the last record, then recovery
-	 * is not needed.
-	 */
-	if ((ret = c->next(c)) == WT_NOTFOUND) {
-		*rec = 0;
+	if ((ret = c->search(c)) == 0) {
+		/*
+		 * If the checkpoint LSN we're given is the last record,
+		 * then recovery is not needed.
+		 */
+		if ((ret = c->next(c)) == WT_NOTFOUND) {
+			*rec = 0;
+			ret = 0;
+		}
+	} else if (ret == WT_NOTFOUND)
+		/*
+		 * If we didn't find that LSN, we need to run recovery,
+		 * but not return any error.
+		 */
 		ret = 0;
-	}
+	else
+		WT_ERR(ret);
 
 err:	WT_TRET(c->close(c));
 	return (ret);
```

However, even after that, and successfully running recovery, the verify was still failing with ENOENT.  The reason is that `__conn_btree_config_set` was failing in the call to `__wt_metadata_search`.  But if I do a `wt -h <dir> list -v` the metadata looks like:
```
metadata:
allocation_size=4KB,app_metadata=,block_allocation=best,block_compressor=,cache_resident=0,checkpoint=(thread-1=(addr="018881e40805b4ce8d81e48bea6a8f8481e4a45004fb808080e2dfc0e301cfc0",order=30,time=1424888568,size=126976,write_gen=60),WiredTigerCheckpoint.35=(addr="018281e46460c7658581e48c76d7ac8b81e4056963c08681e4c63c7418e2cfc0e3021fc0",order=35,time=1424888707,size=147456,write_gen=67)),checkpoint_lsn=(38,128),checksum=uncompressed,collator=,columns=,dictionary=0,format=btree,huffman_key=,huffman_value=,id=0,internal_item_max=0,internal_key_max=0,internal_key_truncate=,internal_page_max=4KB,key_format=S,key_gap=10,leaf_item_max=0,leaf_key_max=0,leaf_page_max=32KB,leaf_value_max=0,memory_page_max=5MB,os_cache_dirty_max=0,os_cache_max=0,prefix_compression=0,prefix_compression_min=4,split_deepen_min_child=0,split_deepen_per_child=0,split_pct=75,value_format=S,version=(major=1,minor=1)
colgroup:wt
app_metadata=,columns=,source="file:wt.wt",type=file
file:wt.wt
allocation_size=512,app_metadata=,block_allocation=best,block_compressor=,cache_resident=0,checkpoint=(thread-1=(addr="01e2b41981e3566e6ee3096f7ec0bee450b66e91e308dbc6c040e4feadaf1a808080e41362f9c0e4112b33c0",order=31,time=1424888568,size=288052224,write_gen=286257),WiredTigerCheckpoint.32=(addr="01e3076d2481e477f2866ce308d66bc010e45977a219e3030b90c088e4fbbbc9c3e308e5aec00fe4e7abd17ee41362f9c0e4111fc5c0",order=32,time=1424888568,size=287303168,write_gen=292252)),checkpoint_lsn=(35,128),checksum="uncompressed",collator=,columns=,dictionary=0,format=btree,huffman_key=,huffman_value=,id=1,internal_item_max=0,internal_key_max=46,internal_key_truncate=false,internal_page_max=2048,key_format=u,key_gap=20,leaf_item_max=0,leaf_key_max=0,leaf_page_max=1024,leaf_value_max=0,memory_page_max=5MB,os_cache_dirty_max=0,os_cache_max=0,prefix_compression=0,prefix_compression_min=6,split_deepen_min_child=0,split_deepen_per_child=0,split_pct=46,value_format=u,version=(major=1,minor=1)
table:wt
app_metadata=,colgroups=,columns=,key_format=u,value_format=u
```
So, the file is listed in the metadata.  The reason for the failure, after a bunch of digging is that the name passed in to `__wt_metadata_search` from `__conn_btree_config_set` is `file:wt` and the metadata cursor is not finding it, even though `file:wt.wt` exists.  So the URI name is not getting normalized for some reason.

If I run `wt -h <dir> verify file:wt.wt` it passes.

I don't know if the end-of-logfile issue is related to the filename issue.  There is another `format-recovery` failure from yesterday that I'll look at.

**Comment by sueloverso** [comment]
Ok, thanks to @keithbostic I was testing by hand with an improper command.  It should have been `table:wt`.  User error on that part.

I tested out my fix above and it solves the failure.  But it does cause recovery to be run, which it should in this crash/abort test anyway.  I'll commit a branch with this workaround.  @michaelcahill we should discuss the use of the `alloc_lsn` as the `ckpt_lsn` in the checkpoint code, since `alloc_lsn` is only a potential LSN, not a guarantee.

If we fixed that use of `alloc_lsn`, then I think it would be a reason to panic if the call in `__wt_log_needs_recovery` is given a checkpoint LSN that does not exist.

**Comment by sueloverso** [comment]
@michaelcahill I am reopening this as I've been thinking about checkpoints and their use of `alloc_lsn` today.  I have a not-yet-fully-debugged set of changes that fixes that issue and, as a bonus, it fixes the problem in #1183.  With this change I can now run recovery in my incremental backup directory and continue to do incremental backups to the directory.  Repeat recovery and incremental backup.  I no longer lose updates as I described in that issue.  I hope to have a branch for perusal soon.
